### PR TITLE
TST: integrate.cumulative_simpson: simplify and strengthen tests

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
scipy/scipy#18151

#### What does this implement/fix?
This replaces the existing N-D tests of `cumulative_simpson` with a property-based test: the results of `cumulative_simpson` with N-D `y` should match the results of applying `cumulative_simpson` to each 1D slice of `y`. The primary intent is to make it easier to verify that all relevant cases are tested; it might also add cases that were not tested before and strengthen the test with the use of random (rather than hand-selected) cases.

I've also attempted to strengthen the 1D tests. I thought the existing test for "exact" (within machine precision) integration of low-order polynomials was a good idea; I went a bit further with this idea.

#### Additional information
Thank you for your patience. It's very important to me that the code I merge into SciPy is correct in its basic behavior, so I would need to write these tests to convince myself of that anyway (whether they end up being committed or not). If you agree that this simplifies and/or strengthens the existing tests, we might as well merge them.

After this, I think we'll be pretty much done!

### CODE REVIEW

File: scipy/integrate/tests/test_quadrature.py [hunk 1]
```diff
@@ -554,166 +584,33 @@ def test_simpson_1d(self, y_func, x, dx, initial, int_y_func):
...
-        y = x**2
-        int_y = x**3 / 3
-        assert_allclose(
-            cumulative_simpson(y, x=input_x, initial=0),
-            int_y,
-        )
+    @pytest.mark.parametrize('axis', np.arange(-3, 3))
+    @pytest.mark.parametrize('x_ndim', (1, 3))
+    @pytest.mark.parametrize('i_ndim', (None, 0, 3,))
+    @pytest.mark.parametrize('dx', (None, True))
```
Comment (mdhaber) [inline scipy/integrate/tests/test_quadrature.py:602]: Parameterize over all combinations of valid `axis`, `x` (None, 1D, and multi-D), `dx` (None, 0D, and multi-D), (None, 0D, and multi-D).

File: scipy/integrate/tests/test_quadrature.py [hunk 2]
```diff
@@ -527,193 +527,97 @@ def func(x):
...
-        x = np.array([input_x]*3)
-        y = x**2
-        int_y = x**3 / 3
-        assert_allclose(
-            cumulative_simpson(y, x=input_x, initial=0),
-            int_y,
-        )
+    @pytest.mark.parametrize('use_dx', (False, True))
+    @pytest.mark.parametrize('use_initial', (False, True))
+    def test_1d(self, use_dx, use_initial):
```
Comment (mdhaber) [inline scipy/integrate/tests/test_quadrature.py:566]: I thought it would be good to test a more general polynomial of the desired order, test a third-order polynomial when the points are equally spaced, and to test all combinations of using `initial` and `dx` vs `x`.

File: scipy/integrate/tests/test_quadrature.py [hunk 3]
```diff
@@ -730,6 +634,18 @@ def test_simpson_exceptions(self, message, kwarg_update):
...
+        # Test special cases not checked elsewhere
+        rng = np.random.default_rng(82456839535679456794)
+        y = rng.random(size=10)
+        res = cumulative_simpson(y, dx=0)
+        assert_equal(res, 0)
+
+        # Should add tests of:
+        # - all elements of `x` identical
+        # - y of size 1 and 2
+        # These should work as they do for `simpson`
```
Comment (mdhaber) [inline scipy/integrate/tests/test_quadrature.py:653]: I put the special case of `dx=0` in a separate test so that it's more visible as such. This brings to mind that there are some other special cases that would ideally return the correct result rather than being rejected by input validation. We don't have to take care of them right now, but it would be a good follow-up.

File: scipy/integrate/tests/test_quadrature.py [hunk 4]
```diff
@@ -527,193 +527,97 @@ def func(x):
...
-    def test_simpson_multid_w_1d_x(self):
-        input_x = [0, 2, 3, 4]
-        x = np.array([input_x]*3)
-        y = x**2
-        int_y = x**3 / 3
-        assert_allclose(
-            cumulative_simpson(y, x=input_x, initial=0),
-            int_y,
-        )
+    @pytest.mark.parametrize('use_dx', (False, True))
```
Comment (nprav) [inline scipy/integrate/tests/test_quadrature.py:564]:
```suggestion
    @pytest.mark.parametrize(('use_dx', 'order'), [(True, 2), (True, 3), (False, 2)])
    @pytest.mark.parametrize('use_initial', (False, True))
    def test_1d(self, use_dx, order, use_initial):
        # Test for exact agreement with polynomial of highest
        # possible order (3 if `dx` is constant, 2 otherwise).
        rng = np.random.default_rng(82456839535679456794)
        n = 10

        # Generate random polynomials and ground truth
        # integral of appropriate order
        order = 3 if use_dx else 2
        dx = rng.random()
        x = (np.sort(rng.random(n)) if order == 2
             else np.arange(n)*dx + rng.random())
        i = np.arange(order + 1)[:, np.newaxis]
        c = rng.random(order + 1)[:, np.newaxis]
        y = np.sum(c*x**i, axis=0)
        Y = np.sum(c*x**(i + 1)/(i + 1), axis=0)

        # Integrate with `cumulative_simpson`
        initial = Y[0] if use_initial else None
        kwarg = {'dx': dx} if use_dx else {'x': x}
        res = cumulative_simpson(y, **kwarg, initial=initial)

        # Compare result against reference
        # If order == 3 and dx is constant, every even value should agree exactly
        # If order == 2, every even value should agree exactly
        if order == 3:
            res = res[::2] if use_initial else res[1::2]
            ref = Y[::2] if use_initial else Y[2::2] - Y[0]
        elif order == 2:
            res = res if use_initial else res[1:]
            ref = Y if use_initial else Y[2:] - Y[0]
        assert_allclose(res, ref)
```

If order == 2, all values should match
If order == 3 and dx = constant, then every other value should match

Comment (mdhaber) [reply scipy/integrate/tests/test_quadrature.py:564]: I know. I chose to keep it simple. I will take another look.

**Comment by mdhaber** [comment]
Otherwise are you happy with this? I think the plan is still to branch tomorrow, so it would be good to merge this tonight.

**Comment by nprav** [comment]
Yup, happy with this. As you said these are simpler but very thorough.
So in total we now have:
- Tests for cumulative_simpson 1d against theory. 
- Tests for simpson nd using 1d as a reference.
- Tests with exceptions
- Tests with an edge case
- Tests comparing to simpson
We should update the 1d test to test all values in the array when order = 2 though. 
Otherwise the intermediate points would not be covered by a test.

**Comment by mdhaber** [comment]
Updated the 1D test.
